### PR TITLE
Added registration views & validation & confirm email

### PR DESCRIPTION
**Pull request by Silent777**
Added activation for user
Added url for token
Added registration template html

## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/secure/RapidBoard.jspa?rapidView=738&view=detail&selectedIssue=LVPYTP-10)

## Code reviewers

- [ ] @dmytrorevak
- [ ] @IvanAndrukh
- [ ] @JjuliaP
- [ ] @makarman
- [ ] @noobhunters
- [ ] @Silent777
- [ ] @Staryj812

### Second Level Review

- [ ] @ssOleg
- [ ] @lhalam

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: eventually/authentication/views.py [hunk 1]
```diff
@@ -2,13 +2,76 @@
...
+            user = CustomUser.objects.get(email=email['email'])
+        except(TypeError, ValueError, OverflowError, CustomUser.DoesNotExist):
+            user = None
+        if user is not None:
+            user.is_active = True
+            user.save()
+            success_msg = 'Thank you for email confirmation. Now you can login your account.'
+            return HttpResponse(success_msg, status=200)
+
+        return HttpResponse('Activation link is invalid!')
```
Comment (Byte4Craft) [inline eventually/authentication/views.py]: If the request method is other than GET, shouldn't we return something in that case?

File: eventually/authentication/urls.py [hunk 2]
```diff
@@ -1,9 +1,10 @@
 from django.conf.urls import url
-from .views import UserView, registration_user, login_user, logout_user
+from .views import UserView, registration, activate, login_user, logout_user
 
 urlpatterns = [
     url(r'^(?P<user_id>[0-9]+)/$', UserView.as_view()),
-    url(r'^register/$', registration_user, name="add_user"),
+    url(r'^register/$', registration, name="add_user"),
```
Comment (redmyt) [inline eventually/authentication/urls.py]: for what do you use the related name for this urls?

File: eventually/authentication/urls.py [hunk 3]
```diff
@@ -1,9 +1,10 @@
...
+    url(r'^activate/(?P<token>.+)$', activate, name='activate'),
```
Comment (redmyt) [inline eventually/authentication/urls.py]: is it necessary to add `.+` after the `<token>`?

File: eventually/authentication/views.py [hunk 4]
```diff
@@ -3,19 +3,66 @@
...
 
 class UserView(View):
     """UserView view handles GET, POST, PUT, DELETE requests."""
 
 
-def registration_user():
+def registration(request):
     """Registration CustomUser"""
 
+    if request.method == 'POST':
```
Comment (redmyt) [inline eventually/authentication/views.py:29]: return something HttpResponse if method != `POST`

File: eventually/authentication/views.py [hunk 5]
```diff
@@ -3,19 +3,66 @@
...
+        data = json_loads(request.body)
```
Comment (redmyt) [inline eventually/authentication/views.py:30]: maybe it makes sense to add the check for `None`?

File: eventually/authentication/views.py [hunk 6]
```diff
@@ -3,19 +3,66 @@
...
+                                 middle_name=data['middle_name'],
+                                 email=data['email'].lower().strip(),
+                                 password=data['password'])
+        if not user:
+            return HttpResponse('Email is already exist', status=400)
+
+        ctx = {
+            'first_name': user.first_name,
+            'domain': FRONT_HOST,
+            'token': create_token({'email': user.email}),
```
Comment (redmyt) [inline eventually/authentication/views.py]: add expiration time for the token

File: eventually/authentication/views.py [hunk 7]
```diff
@@ -3,19 +3,66 @@
...
+        mail_subject = 'Activate account'
+        send_email(mail_subject, message, [user.email], 'registration.html', ctx)
+        msg = 'Please confirm your email address to complete the registration'
+        return HttpResponse(msg, status=201)
+
+
+def activate(request, token):
+    """Activation CustomUser"""
+    if request.method == 'GET':
+        try:
```
Comment (redmyt) [inline eventually/authentication/views.py]: it is not good practice to use `try` in the views. Choose other function than `objects.get`

File: eventually/authentication/views.py [hunk 8]
```diff
@@ -3,19 +3,66 @@
...
+            email = handle_token(token)
+            user = CustomUser.objects.get(email=email['email'])
+        except(TypeError, ValueError, OverflowError, CustomUser.DoesNotExist):
+            user = None
+        if user is not None:
```
Comment (redmyt) [inline eventually/authentication/views.py]: why not only `if user`?

File: eventually/authentication/views.py [hunk 9]
```diff
@@ -3,19 +3,66 @@
...
+            user.is_active = True
```
Comment (redmyt) [inline eventually/authentication/views.py]: better use the model's update method and remove `save()`

File: eventually/eventually/settings.py [hunk 10]
```diff
@@ -139,3 +139,9 @@
     from .local_settings import *
 except ImportError:
     pass
+
+JWT_TOKEN_KEY = 'any secret word'
```
Comment (redmyt) [inline eventually/eventually/settings.py]: this is secret info, store it in the local_settings and remove from here.

File: eventually/eventually/settings.py [hunk 11]
```diff
@@ -139,3 +139,9 @@
...
+
+JWT_ALGORITHM = 'HS256'
```
Comment (redmyt) [inline eventually/eventually/settings.py]: this is secret info, store it in the local_settings and remove from here.

File: eventually/utils/validators.py [hunk 12]
```diff
@@ -120,3 +123,42 @@ def list_of_int_validator(value):
     if not all(isinstance(item, int) for item in value):
         return False
     return True
+
+
+def email_validator(email):
+    """
+    Function that provides string validation.
```
Comment (redmyt) [inline eventually/utils/validators.py]: provides email validation not string

File: eventually/utils/validators.py [hunk 13]
```diff
@@ -120,3 +123,42 @@ def list_of_int_validator(value):
...
+
+    :param email: String with email data
+    :type email: str
+
+    :return: `True` if email if valid and `False` if it is not.
+    """
+
+    email = str(email).lower().strip()
```
Comment (redmyt) [inline eventually/utils/validators.py]: do we need this transformation in the validation function?

File: eventually/utils/validators.py [hunk 14]
```diff
@@ -120,3 +123,42 @@ def list_of_int_validator(value):
...
+
+    try:
+        validate_email(email)
+        return True
+    except ValidationError:
+        return False
```
Comment (redmyt) [inline eventually/utils/validators.py]: maybe `pass`?

File: eventually/utils/validators.py [hunk 15]
```diff
@@ -120,3 +123,42 @@ def list_of_int_validator(value):
...
+
+
+def registration_validate(data):
+    """
+
+    :param data:
+    :return:
+    """
+    if not email_validator(data['email']):
```
Comment (redmyt) [inline eventually/utils/validators.py]: raise error when data has not attribute 'email'

File: eventually/utils/validators.py [hunk 16]
```diff
@@ -120,3 +123,42 @@ def list_of_int_validator(value):
...
+        return False
+    required_keys = ['email', 'password']
+    if not required_keys_validator(data, required_keys, strict=False):
+        return False
+    if not string_validator(data['first_name'], 1, 20):
```
Comment (redmyt) [inline eventually/utils/validators.py]: raise error when data has not attribute 'first name'

File: eventually/utils/validators.py [hunk 17]
```diff
@@ -120,3 +123,42 @@ def list_of_int_validator(value):
...
+        return False
+    if not string_validator(data['middle_name'], 1, 20):
```
Comment (redmyt) [inline eventually/utils/validators.py]: raise error when data has not attribute 'middle name'

File: eventually/utils/validators.py [hunk 18]
```diff
@@ -120,3 +123,42 @@ def list_of_int_validator(value):
...
+        return False
+    if not string_validator(data['last_name'], 1, 20):
```
Comment (redmyt) [inline eventually/utils/validators.py]: raise error when data has not attribute 'last name'

File: eventually/authentication/views.py (hunk 4 above)
Comment (Silent777) [reply eventually/authentication/views.py:29]: Dear friend.
I`ll return HttpResponse(status=400) 
best regards,
Oleksii :)

File: eventually/authentication/views.py (hunk 5 above)
Comment (Silent777) [reply eventually/authentication/views.py:30]: ++

File: eventually/authentication/views.py (hunk 1 above)
Comment (Silent777) [reply eventually/authentication/views.py]: Dear friend.
I`ll return HttpResponse(status=400) 
best regards,
Oleksii :)

File: eventually/utils/validators.py (hunk 14 above)
Comment (Silent777) [reply eventually/utils/validators.py]: Dear friend.
I will consider your wishes
best regards,
Oleksii :)

**Review by lhalam — APPROVED** [review]

**Pull request merged**